hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconIncrementalContainerReportHandler.java

```diff
@@ -70,19 +70,22 @@ public void onMessage(final IncrementalContainerReportFromDatanode report,
         final DatanodeDetails dd = report.getDatanodeDetails();
         final ContainerID id = ContainerID.valueof(
             replicaProto.getContainerID());
-        if (!getContainerManager().exists(id)) {
-          LOG.info("New container {} got from {}.", id,
-              report.getDatanodeDetails());
-          try {
-            ContainerWithPipeline containerWithPipeline =
-                scmClient.getContainerWithPipeline(id.getId());
-            LOG.info("Verified new container from SCM {} ",
-                containerWithPipeline.getContainerInfo().containerID());
-            containerManager.addNewContainer(id.getId(), containerWithPipeline);
-          } catch (IOException ioEx) {
-            LOG.error("Exception while getting new container info from SCM",
-                ioEx);
-            return;
+        synchronized (containerManager) {
+          if (!containerManager.exists(id)) {
+            LOG.info("New container {} got from {}.", id,
+                report.getDatanodeDetails());
+            try {
+              ContainerWithPipeline containerWithPipeline =
+                  scmClient.getContainerWithPipeline(id.getId());
+              LOG.info("Verified new container from SCM {} ",
+                  containerWithPipeline.getContainerInfo().containerID());
+              containerManager
+                  .addNewContainer(id.getId(), containerWithPipeline);
+            } catch (IOException ioEx) {
+              LOG.error("Exception while getting new container info from SCM",
+                  ioEx);
+              return;
+            }
```

Comment:
Looks like this block could be shared between the two container report handlers.